lib/Frontend/CompilerInvocation.cpp

```diff
@@ -222,26 +222,161 @@ class ArgsToFrontendInputsConverter {
     }
     return false;
   }
+};
+    class FrontendArgsToOptionsConverter {
+    private:
+        DiagnosticEngine &Diags;
+        const llvm::opt::ArgList &Args;
+        FrontendOptions &Opts;
+        
+        llvm::Optional<const std::vector<std::string>>
+        cachedOutputFilenamesFromCommandLineOrFilelist;
+        
+
+
+  // This is a separate function so that it shows up in stack traces.
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithAssertion() {
+    // This assertion should always fail, per the user's request, and should
+    // not be converted to llvm_unreachable.
+    assert(0 && "This is an assertion!");
+  }
+
+  // This is a separate function so that it shows up in stack traces.
+  LLVM_ATTRIBUTE_NOINLINE
+  static void debugFailWithCrash() { LLVM_BUILTIN_TRAP; }
+
+  void computeDebugCrashGroup();
+  void computeDebugTimeOptions();
+  void computePrintStatsOptions();
+  void computeTBDOptions();
+  void setUnsignedIntegerArgument(options::ID optionID, unsigned max,
+                                  unsigned &valueToSet);
+  void computePlaygroundOptions();
+  void computeHelpOptions();
+  void computeDumpScopeMapLocations();
+  FrontendOptions::ActionType determineRequestedAction() const;
+  bool setupForSILOrLLVM();
+  bool computeModuleName();
+  bool computeFallbackModuleName();
+  bool computeOutputFilenames();
+  bool deriveOutputFilenameFromInputFile();
+  bool deriveOutputFilenameForDirectory(llvm::StringRef outputDir);
+  std::string computeBaseNameOfOutput() const;
+  void determineSupplementaryOutputFilenames();
+  /// Returns the output filenames on the command line or in the output
+  /// filelist. If there was an error (reading the list) returns None. If there
+  /// were neither -o's nor an output filelist, returns an empty vector.
+  const llvm::Optional<const std::vector<std::string>> &
+  getOutputFilenamesFromCommandLineOrFilelist();
+  bool hasAnUnusedOutputPath() const;
+  void computeImportObjCHeaderOptions();
+  void computeImplicitImportModuleNames();
+  void computeLLVMArgs();
+  const std::vector<std::string>
+  readOutputFileList(const StringRef filelistPath) const;
```

Comment:
…and isn't really useful on by-value arguments either. This one's a bit of a conflict because it *does* mean something in the *body* of the function, but it's distracting when reading the function signature.

Reply:
I would like to push back. It guarantees that the body cannot contain "filelistPath = <foo>".
In this case, it's not a big deal because the function body is short, but I would like to cultivate the habit.